contracts/chain-adapters/Solana_Adapter.sol

```diff
@@ -0,0 +1,193 @@
+    }
+
+    /**
+     * @notice Send cross-chain message to target on Solana.
+     * @dev Only allows sending messages to the Solana spoke pool.
+     * @param target Program on Solana (translated as EVM address) that will receive message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        if (target != SOLANA_SPOKE_POOL_ADDRESS) {
+            revert InvalidRelayMessageTarget(target);
+        }
+
+        bytes4 selector = bytes4(message[:4]);
+        if (selector == SpokePoolInterface.setEnableRoute.selector) {
+            cctpMessageTransmitter.sendMessage(
+                CircleDomainIds.Solana,
+                SOLANA_SPOKE_POOL_BYTES32,
+                _translateSetEnableRoute(message)
+            );
+        } else {
+            cctpMessageTransmitter.sendMessage(CircleDomainIds.Solana, SOLANA_SPOKE_POOL_BYTES32, message);
+        }
+
+        // TODO: consider if we need also to emit the translated message.
+        emit MessageRelayed(target, message);
+    }
+
+    /**
+     * @notice Bridge tokens to Solana.
+     * @dev Only allows bridging USDC to Solana spoke pool.
+     * @param l1Token L1 token to deposit.
+     * @param l2Token L2 token to receive.
+     * @param amount Amount of L1 tokens to deposit and L2 tokens to receive.
+     * @param to Bridge recipient.
+     */
+    function relayTokens(
+        address l1Token,
+        address l2Token,
+        uint256 amount,
+        address to
+    ) external payable override {
+        if (l1Token != address(usdcToken)) {
+            revert InvalidL1Token(l1Token);
+        }
+        if (l2Token != SOLANA_USDC_ADDRESS) {
+            revert InvalidL2Token(l2Token);
+        }
+        if (amount > type(uint64).max) {
+            revert InvalidAmount(amount);
+        }
+        if (to != SOLANA_SPOKE_POOL_ADDRESS) {
+            revert InvalidTokenRecipient(to);
+        }
+
+        _transferUsdc(SOLANA_SPOKE_POOL_USDC_VAULT, amount);
+
+        // TODO: consider if we need also to emit the translated addresses.
+        emit TokensRelayed(l1Token, l2Token, amount, to);
+    }
+
+    /**
+     * @notice Helper to map a Solana address to an Ethereum address representation.
+     * @dev The Ethereum address is derived from the Solana address by hashing it and then truncating to its lowest 20
+     * bytes. This same conversion must be done by the HubPool owner when adding Solana spoke pool and setting the
+     * corresponding pool rebalance and deposit routes.
+     * @param solanaAddress Solana address (Base58 decoded to bytes32) to map to its Ethereum address representation.
+     * @return Ethereum address representation of the Solana address.
+     */
+    function _mapSolanaAddress(bytes32 solanaAddress) internal pure returns (address) {
+        return address(uint160(uint256(keccak256(abi.encodePacked(solanaAddress)))));
```

Comment:
Why do you need to hash the bytes32 string?

Reply:
Initially I added hash just to be on the safe side to avoid collisions, but we only here have couple of addresses and they already should be pretty homogeneous. The only case where this would matter is native system programs that have zeros in their trailing bytes. But we need to represent only key or seed derived addresses on EVM, so their bytes32 representation should be pretty homogeneous.

Pushed a commit to drop the hashing and only trim off the first 12 bytes.